File: aide/cmd/aide/gitdir_shapes_test.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// gitdirShapeCase mirrors testdata/gitdir-shapes.json — the golden table
// shared with src/test/gitdir-shapes.test.ts so the Go resolver and the TS
// mirror cannot drift on gitdir classification.
type gitdirShapeCase struct {
	Name         string   `json:"name"`
	GitdirDirs   []string `json:"gitdirDirs"`
	GitdirTarget *string  `json:"gitdirTarget"`
	CwdSuffix    string   `json:"cwdSuffix"`
	ExpectAnchor string   `json:"expectAnchor"`
	ExpectShape  string   `json:"expectShape"`
}

func loadGitdirShapeCases(t *testing.T) []gitdirShapeCase {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "..", "..", "testdata", "gitdir-shapes.json"))
	if err != nil {
		t.Fatalf("golden table: %v", err)
	}
	var table struct {
		Cases []gitdirShapeCase `json:"cases"`
	}
	if err := json.Unmarshal(data, &table); err != nil {
		t.Fatalf("golden table: %v", err)
	}
	if len(table.Cases) == 0 {
		t.Fatal("golden table is empty")
	}
	return table.Cases
}

func TestGitdirShapesGolden(t *testing.T) {
	t.Setenv("AIDE_PROJECT_ROOT", "")
	os.Unsetenv("AIDE_PROJECT_ROOT")

	for _, tc := range loadGitdirShapeCases(t) {
		t.Run(tc.Name, func(t *testing.T) {
			tmp, err := filepath.EvalSymlinks(t.TempDir())
			if err != nil {
				t.Fatal(err)
			}
			super := filepath.Join(tmp, "super")
			for _, d := range tc.GitdirDirs {
				if err := os.MkdirAll(filepath.Join(super, filepath.FromSlash(d)), 0o755); err != nil {
					t.Fatal(err)
				}
			}
			checkout := filepath.Join(tmp, "checkout")
			cwd := checkout
			if tc.CwdSuffix != "" {
				cwd = filepath.Join(checkout, filepath.FromSlash(tc.CwdSuffix))
			}
			if err := os.MkdirAll(cwd, 0o755); err != nil {
				t.Fatal(err)
			}

			gitPath := filepath.Join(checkout, ".git")
			switch {
			case tc.GitdirTarget == nil:
				if err := os.MkdirAll(gitPath, 0o755); err != nil {
					t.Fatal(err)
				}
			case *tc.GitdirTarget == "GARBAGE":
				if err := os.WriteFile(gitPath, []byte("not a gitdir pointer\n"), 0o644); err != nil {
					t.Fatal(err)
				}
			default:
				content := "gitdir: " + filepath.Join(super, filepath.FromSlash(*tc.GitdirTarget)) + "\n"
				if err := os.WriteFile(gitPath, []byte(content), 0o644); err != nil {
					t.Fatal(err)
				}
			}

			a := resolveAnchor(cwd)

			wantRoot := checkout
			if tc.ExpectAnchor == "super" {
				wantRoot = super
			}
			if a.Root != wantRoot {
				t.Errorf("anchor = %q, want %s (%q)", a.Root, tc.ExpectAnchor, wantRoot)
			}
			if a.Provenance.GitdirShape != tc.ExpectShape {
				t.Errorf("gitdirShape = %q, want %q", a.Provenance.GitdirShape, tc.ExpectShape)
			}
		})
	}
}

File: src/test/gitdir-shapes.test.ts
/**
 * Golden gitdir-shape table (testdata/gitdir-shapes.json) run against the
 * TS resolver mirror — the same table the Go side runs in
 * aide/cmd/aide/gitdir_shapes_test.go, so the two implementations cannot
 * drift on gitdir classification. TS asserts anchoring only (the TS
 * resolver exposes no provenance surface).
 *
 * Run with: npx vitest run src/test/gitdir-shapes.test.ts
 */

import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import {
  mkdtempSync,
  rmSync,
  mkdirSync,
  writeFileSync,
  readFileSync,
  realpathSync,
} from "fs";
import { join, dirname } from "path";
import { tmpdir } from "os";
import { fileURLToPath } from "url";

let tempHome = "";

vi.mock("os", async (importOriginal) => {
  const actual = (await importOriginal()) as typeof import("os");
  return {
    ...actual,
    homedir: () => tempHome,
  };
});

interface ShapeCase {
  name: string;
  gitdirDirs: string[];
  gitdirTarget: string | null;
  cwdSuffix?: string;
  expectAnchor: "checkout" | "super";
}

const tableFile = join(
  dirname(fileURLToPath(import.meta.url)),
  "..",
  "..",
  "testdata",
  "gitdir-shapes.json",
);
const cases = (
  JSON.parse(readFileSync(tableFile, "utf-8")) as { cases: ShapeCase[] }
).cases;

describe("gitdir shapes golden table", () => {
  let tmp: string;

  beforeEach(() => {
    tmp = realpathSync(mkdtempSync(join(tmpdir(), "aide-shapes-")));
    tempHome = realpathSync(mkdtempSync(join(tmpdir(), "aide-shapes-home-")));
    delete process.env.AIDE_PROJECT_ROOT;
    vi.resetModules();
  });

  afterEach(() => {
    rmSync(tmp, { recursive: true, force: true });
    rmSync(tempHome, { recursive: true, force: true });
    delete process.env.AIDE_PROJECT_ROOT;
  });

  for (const tc of cases) {
    it(tc.name, async () => {
      const { findProjectRoot } = await import("../lib/project-root.js");

      const superRepo = join(tmp, "super");
      for (const d of tc.gitdirDirs) {
        mkdirSync(join(superRepo, ...d.split("/")), { recursive: true });
      }
      const checkout = join(tmp, "checkout");
      const cwd = tc.cwdSuffix
        ? join(checkout, ...tc.cwdSuffix.split("/"))
        : checkout;
      mkdirSync(cwd, { recursive: true });

      const gitPath = join(checkout, ".git");
      if (tc.gitdirTarget === null) {
        mkdirSync(gitPath, { recursive: true });
      } else if (tc.gitdirTarget === "GARBAGE") {
        writeFileSync(gitPath, "not a gitdir pointer\n");
      } else {
        writeFileSync(
          gitPath,
          `gitdir: ${join(superRepo, ...tc.gitdirTarget.split("/"))}\n`,
        );
      }

      const result = findProjectRoot(cwd);
      const wantRoot = tc.expectAnchor === "super" ? superRepo : checkout;
      expect(result.root).toBe(wantRoot);
      expect(result.hasMarker).toBe(true);
    });
  }
});
